diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/SmartCastRule.kt

```diff
@@ -0,0 +1,152 @@
+import org.cqfn.diktat.ruleset.utils.getAllChildrenWithType
+import org.cqfn.diktat.ruleset.utils.getFirstChildWithType
+import org.cqfn.diktat.ruleset.utils.hasChildOfType
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+
+class SmartCastRule(private val configRules: List<RulesConfig>) : Rule("smart-cast-rule") {
+
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == IF) {
+            handleIfBlock(node)
+        }
+
+        if (node.elementType == WHEN) {
+            handleWhenCondition(node)
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleIfBlock(node: ASTNode) {
+        /*
+            There are 2 conditions.
+            1) if has 'is' in condition, then we check all 'as' keywords in 'then block'
+            2) if has '!is' -> NOT_IS in condition, then we check all 'as' keywords in else's 'then block'
+         */
+        val isBlocks = mutableListOf<IsExpressions>()
+        val notIsBlocks = mutableListOf<IsExpressions>()
+
+        val conditionBlock = node.findAllNodesWithSpecificType(CONDITION).first()
+
+        conditionBlock.findAllNodesWithSpecificType(IS_EXPRESSION).map { it.text }.forEach {
+            if (it.contains("!is")) {
+                val split = it.split("!is").map { part -> part.trim() }
+                notIsBlocks.add(IsExpressions(split[0], split[1]))
+            } else {
+                val split = it.split("is").map { part -> part.trim() }
+                isBlocks.add(IsExpressions(split[0], split[1]))
+            }
+        }
+
+        val elseBlock = node.findAllNodesWithSpecificType(ELSE).firstOrNull { it.firstChildNode.elementType != IF }
+
+        if (isBlocks.isNotEmpty()) {
+            val then = node.findChildByType(THEN)!!
+            handleThenBlock(then, isBlocks)
+        }
+
+        if (notIsBlocks.isNotEmpty() && elseBlock != null) {
+            handleThenBlock(elseBlock, notIsBlocks)
+        }
+    }
+
+    private fun handleThenBlock(then: ASTNode, blocks: List<IsExpressions>) {
+        val thenBlock = then.findChildByType(BLOCK)
+
+        if (thenBlock != null) {
+            // Find all as expressions that are inside this current block
+            val asList = thenBlock.findAllNodesWithSpecificType(BINARY_WITH_TYPE).filter { it.text.contains(" as ")
+                    && it.findParentNodeWithSpecificType(BLOCK) == thenBlock }
+            checkAsExpressions(asList, blocks)
+        } else {
+            val asList = then.findAllNodesWithSpecificType(BINARY_WITH_TYPE).filter {  it.text.contains(" as ") }
```

Comment:
You should check here that variable from condition is not shadowed. You can take a look at `LocalVariableRule` and `PsiUtils` for some examples of how to do similar checks.